File: widgetssdk/src/main/java/com/glia/widgets/core/authentication/Authentication.java
package com.glia.widgets.core.authentication;

import androidx.annotation.NonNull;

import com.glia.androidsdk.RequestCallback;
import com.glia.widgets.di.Dependencies;

public class Authentication implements com.glia.androidsdk.visitor.Authentication {
    private final com.glia.androidsdk.visitor.Authentication authentication;

    public Authentication(com.glia.androidsdk.visitor.Authentication authentication) {
        this.authentication = authentication;
    }

    @Override
    public void setBehavior(@NonNull Behavior behavior) {
        authentication.setBehavior(behavior);
    }

    @Override
    public void authenticate(RequestCallback<Void> requestCallback, String jwtToken) {
        Dependencies.getControllerFactory().destroyControllers();
        authentication.authenticate(requestCallback, jwtToken);
    }

    @Override
    public void deauthenticate(RequestCallback<Void> requestCallback) {
        Dependencies.getControllerFactory().destroyControllers();
        authentication.deauthenticate(requestCallback);
    }

    @Override
    public boolean isAuthenticated() {
        return authentication.isAuthenticated();
    }
}

File: widgetssdk/src/main/java/com/glia/widgets/core/authentication/Authentication.kt
package com.glia.widgets.core.authentication

import com.glia.androidsdk.RequestCallback
import com.glia.androidsdk.visitor.Authentication
import com.glia.widgets.di.Dependencies

/**
 * Wrapper class for {@link com.glia.androidsdk.visitor.Authentication}
 * Its purpose is to execute Widgets-specific code (e.g. destroy controllers)
 *
 */
class Authentication(private val authentication: Authentication) : Authentication {
    override fun setBehavior(behavior: Authentication.Behavior) {
        authentication.setBehavior(behavior)
    }

    override fun authenticate(requestCallback: RequestCallback<Void>, jwtToken: String) {
        Dependencies.getControllerFactory().destroyControllers()
        authentication.authenticate(requestCallback, jwtToken)
    }

    override fun deauthenticate(requestCallback: RequestCallback<Void>) {
        Dependencies.getControllerFactory().destroyControllers()
        authentication.deauthenticate(requestCallback)
    }

    override fun isAuthenticated(): Boolean {
        return authentication.isAuthenticated
    }
}
